identity_iota/src/tangle_writer.rs

```diff
@@ -0,0 +1,174 @@
+use crate::helpers::get_iota_address;
+use async_trait::async_trait;
+pub use identity_core::did::IdentityWriter;
+use identity_core::did::{DIDDocument, DID};
+pub use iota::client::builder::Network as iota_network;
+use iota::{
+    client::Transfer,
+    crypto::ternary::{
+        sponge::{CurlP81, Sponge},
+        Hash,
+    },
+    ternary::{T1B1Buf, TritBuf, TryteBuf},
+    transaction::bundled::{Address, BundledTransaction, BundledTransactionField, Tag},
+};
+use iota_conversion::Trinary;
+use serde::{Deserialize, Serialize};
+use std::{fmt, thread, time::Duration};
+
+#[async_trait]
+impl IdentityWriter for IOTAWriter {
+    type Payload = Payload;
+    type Hash = Hash;
+    type Error = crate::Error;
+    async fn send(&self, did_document: &Payload) -> crate::Result<Hash> {
+        let mut tail_transaction = self.publish_document(did_document).await?;
+        thread::sleep(Duration::from_secs(5));
+        let mut j: usize = 0;
+        while !self.is_confirmed(tail_transaction).await? {
+            j += 1;
+            thread::sleep(Duration::from_secs(5));
+            self.promote(tail_transaction).await?;
+            // Send the document again if the previous transaction didn't get confirmed after 150 seconds
+            if j % 30 == 0 {
+                tail_transaction = self.publish_document(&did_document).await?;
+            }
+        }
+        Ok(tail_transaction)
+    }
+}
+
+pub struct IOTAWriter {
+    iota: iota::Client,
+    network: iota::client::builder::Network,
+}
+
+impl IOTAWriter {
+    pub async fn new(nodes: Vec<&'static str>, network: iota::client::builder::Network) -> crate::Result<IOTAWriter> {
+        let iota = iota::ClientBuilder::new()
+            .nodes(&nodes)?
+            .network(network.clone())
+            .build()?;
+        Ok(Self { iota, network })
+    }
+    /// Publishes DID document or diff to the Tangle
+    pub async fn publish_document(&self, did_document: &Payload) -> crate::Result<Hash> {
+        let (did, document_string) = match did_document {
+            Payload::DIDDocument(document) => (document.derive_did().clone(), document.to_string()),
+            Payload::DIDDocumentDifferences(differences) => {
+                (differences.did.clone(), serde_json::to_string(&differences)?)
+            }
+        };
+        // Check if correct network selected
+        check_network(did.id_segments.clone(), &self.network)?;
+
+        // Where does the address for diff changes come from?
+        let address = get_iota_address(&did)?;
+
+        let transfers = vec![Transfer {
+            address: Address::from_inner_unchecked(TryteBuf::try_from_str(&address)?.as_trits().encode()),
+            value: 0,
+            message: Some(document_string),
+            tag: Some(
+                Tag::try_from_inner(
+                    TryteBuf::try_from_str("DID999999999999999999999999")?
```

Comment:
It's the tag for the transactions so you can find all DID transactions with it, but it's not required for functionality, theoretically a user can provide it, but it will be removed with chrysalis pt. 2 anyways